viite-backend/viite-main/src/main/scala/fi/vaylavirasto/viite/dynamicnetwork/DynamicRoadNetworkService.scala

```diff
@@ -0,0 +1,564 @@
+        val newEndM = change.newEndM
+        val changesWithNewLinkId = tiekamuRoadLinkChanges.filter(ch => ch.newLinkId == newLinkId)
+        val otherChangesWithSameNewLinkId = changesWithNewLinkId.filter(ch => ch.newStartM != newStartM && ch.newEndM != newEndM)
+
+        val newlinkIdChangesLength = GeometryUtils.scaleToThreeDigits(changesWithNewLinkId.map(ch => ch.newEndM).max - changesWithNewLinkId.map(ch => ch.newStartM).min)
+        val newLinkKGV = kgvRoadLinks.find(kgvLink => kgvLink.linkId == newLinkId).getOrElse(throw ViiteException(s"Missing new link from kgv. Cannot validate Tiekamu change infos without kgv road link. LinkId: ${newLinkId}, changeInfo: ${change}"))
+
+        val linearLocationsWithOldLinkId = linearLocations.filter(_.linkId == oldLinkId)
+        val roadAddressedLinkLength = GeometryUtils.scaleToThreeDigits(linearLocationsWithOldLinkId.map(_.endMValue).max - linearLocationsWithOldLinkId.map(_.startMValue).min)
+
+        // check that there are no "partial" changes to road addressed links, i.e. only part of the link changes and the other part has no changes applied to it.
+        if (lengthOfChange != roadAddressedLinkLength) {
+          val allChangesWithOldLinkId = tiekamuRoadLinkChanges.filter(_.oldLinkId == oldLinkId)
+          val combinedLengthOfChanges = GeometryUtils.scaleToThreeDigits(allChangesWithOldLinkId.map(och => och.oldEndM - och.oldStartM).sum)
+          if (combinedLengthOfChanges != roadAddressedLinkLength)
+            tiekamuRoadLinkChangeErrors += TiekamuRoadLinkChangeError("No partial changes allowed. The old link needs to have changes applied to the whole length of the old link", change, getMetaData(change, linearLocations))
+        }
+        else if (newlinkIdChangesLength != GeometryUtils.scaleToThreeDigits(newLinkKGV.length)) {
+          tiekamuRoadLinkChangeErrors += TiekamuRoadLinkChangeError("No partial changes allowed. The new link needs to have changes applied to the whole length of the new link", change, getMetaData(change, linearLocations))
+        }
+
+        // if there are combined links (A + B = C)
+        else if (otherChangesWithSameNewLinkId.nonEmpty) {
+          val oldLinkIds = otherChangesWithSameNewLinkId.map(_.oldLinkId)
+          val oldLinearLocations = linearLocations.filter(ll => oldLinkIds.contains(ll.linkId))
+          val roadways = roadwayDAO.fetchAllByRoadwayNumbers(oldLinearLocations.map(_.roadwayNumber).toSet)
+          // group the roadways with roadNumber, roadPartNumber and track
+          val roadGroups = roadways.groupBy(rw => (rw.roadNumber, rw.roadPartNumber, rw.track))
+          // if there are change infos that combine two or more links but the road address is not homogeneous between those merging links then its an error
+          val isRoadAddressHomogeneous = checkRoadAddressHomogeneity(roadGroups)
+          val changesInSameRoadway = roadGroups.head._2.map(_.roadwayNumber).distinct.size == 1
```

Comment:
Tälle riville joku kommentti lukemista nopeuttamaan / ymmärrystä helpottamaan
// used only inside "if (isRoadAddressHomogeneous)", when we already know roadGroup has only one item, thus handling .head only is enough

...tosin, en tiä onks toi kommentti sit taas tarpeen, jos toi  roadGroups.size < 2 on heti näkyvillä tos ylemmällä rivillä jo.